### PR TITLE
[FEAT] Implement user authentication features including registration, login, and JWT token management

### LINKED ISSUE #64: Base User Model + Authentication
## Files:
	•	app/models/users/user.py
	•	app/schemas/user.py
	•	app/services/auth_service.py

## Tasks:
	•	Implement User model with basic fields (id, email, password_hash, user_type, created_at)
	•	Add UserType enum (FARMER, BUYER, ADMIN)
	•	Create Pydantic schemas: UserCreate, UserResponse, UserLogin
	•	Implement AuthService with basic login/register logic
	•	Add password hashing using bcrypt

**✅ Test Criteria**:
```bash
# Should work:
curl -X POST localhost:8000/auth/register -d '{"email":"test@test.com","password":"123","user_type":"FARMER"}'
curl -X POST localhost:8000/auth/login -d '{"email":"test@test.com","password":"123"}'
make test  # Tests should pass

### PR DESCRIPTION
**Pull request by WassCodeur**
# Pull Request RevolutionaryFarmers

## 📝 Summary

Added base user model and basic auth logic (register/login) with password hashing and JSON API support.

Implemented User SQLAlchemy model with fields: id, email, password_hash, user_type, created_at, etc.
Created UserType enum with values: FARMER, BUYER, ADMIN.
Added Pydantic schemas: UserCreate, UserLogin, and UserResponse.
Implemented AuthService with register and login logic.
Added password hashing via bcrypt.

---

### 🪧 Related Issues

- Closes #64 

---

### 🏁 Type of Change

Mark with an `x` all the checkboxes that apply (like `[x]`).

- [x] 📝 Documentation (updates to README, docs, or comments)
- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 👌 Enhancement (non-breaking change which adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to change)

---

### 🔄 Changes Made

Provide a clear and concise description of what you changed and why.

---

### 🚀 Implementation Details

Describe how you implemented the changes.

---

### 🛠 Technical Notes

Include any technical details that reviewers should be aware of.

---

### ✅ Tests Results

Describe the tests you performed to verify your changes:

Adjusted tests to send JSON data with correct Content-Type: application/json header.

Example working commands:

```bash

curl -X POST localhost:8000/auth/register \
  -H "Content-Type: application/json" \
  -d '{"email":"test@test.com","password":"123","user_type":"FARMER"}'

curl -X POST localhost:8000/auth/login \
  -H "Content-Type: application/json" \
  -d '{"email":"test@test.com","password":"123"}'
```

---

### Test Coverage

- [ ] ✅ Unit Tests
- [ ] ✅ Integration Tests
- [x] ✅ Manual Testing

---

### 📸 Evidence

Screenshot from 2025-07-02 17-52-03

---

### 🔍 Testing Notes

Include any special testing considerations or edge cases checked.

---

### 🔜 Next Steps

Indicate actions or improvements to be taken after this PR, if applicable.

---

### CODE REVIEW

File: app/services/auth_service.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+from fastapi import HTTPException, status
+from jose import JWTError, jwt
+from passlib.context import CryptContext
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy import select
+
+from app.core.config import get_settings
+from app.models.users.user import User
+from app.schemas.user import UserCreate, UserInDB, TokenData
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/services/auth_service.py:15]: @WassCodeur Remove import UserInDB

File: app/models/users/user.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    __tablename__ = "users"
+    
+    # Override id to use UUID
+    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
+    email = Column(String, unique=True, index=True, nullable=False)
+    password_hash = Column(String, nullable=False)
+    user_type = Column(
+    Enum(UserType, name="usertype", create_type=False),
+    nullable=False)
+    is_active = Column(Boolean, default=True)
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/models/users/user.py:31]: Fix sintax in line code 29 and 30

File: app/models/users/user.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+"""
+User model for authentication and user management.
+"""
+import enum
+from sqlalchemy import Column, String, Enum, DateTime, Boolean, func
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/models/users/user.py:5]: Remove imports DateTime and func

File: app/main.py [hunk 4]
```diff
@@ -52,8 +53,9 @@ async def lifespan(app: FastAPI):
     allow_headers=["*"],
 )
 
-# Include GraphQL router
+# Include routers
 app.include_router(graphql_router, prefix="/graphql", tags=["graphql"])
+app.include_router(auth_router)
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/main.py:58]: Please add the prefix and tag

@WassCodeur

File: app/services/auth_service.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+        
+        try:
+            payload = jwt.decode(
+                token, 
+                self.settings.secret_key, 
+                algorithms=[self.settings.algorithm]
+            )
+            email: str = payload.get("sub")
```
Comment (aguilar1x) [inline app/services/auth_service.py:102]: sintax error in payload.get("sub"

```
Type "Any | None" is not assignable to declared type "str"
  Type "Any | None" is not assignable to type "str"
    "None" is not assignable to "str"basedpyright[reportAssignmentType](https://docs.basedpyright.com/v1.29.4/configuration/config-files/#reportAssignmentType)
  ```

File: app/services/auth_service.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            if email is None:
+                raise credentials_exception
+            token_data = TokenData(email=email)
+        except JWTError:
+            raise credentials_exception
+        
+        user = await self.get_user_by_email(db, email=token_data.email)
```
Comment (aguilar1x) [inline app/services/auth_service.py:109]: sintax error in token_data.email

```bash
Argument of type "str | None" cannot be assigned to parameter "email" of type "str" in function "get_user_by_email"
  Type "str | None" is not assignable to type "str"
    "None" is not assignable to "str"basedpyright[reportArgumentType](https://docs.basedpyright.com/v1.29.4/configuration/config-files/#reportArgumentType)
```

File: app/schemas/user.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+class UserBase(BaseModel):
+    """Base user schema with common fields."""
+    email: EmailStr
+    user_type: UserType
+
+
+class UserCreate(UserBase):
+    """Schema for user creation."""
+    password: str = Field(..., min_length=3, description="Password must be at least 3 characters") # noqa
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/schemas/user.py:21]: no lo dejes de 3, dejalo en un tamaño 6 u 8

@WassCodeur

File: app/services/__init__.py [hunk 8]
```diff
@@ -11,11 +10,12 @@
...
-# from .farmer_service import FarmerService
-# from .product_service import ProductService
+# TODO: Import schemas as they are implemented
+# from .farmer import FarmerCreate, FarmerResponse
+# from .product import ProductCreate, ProductResponse
+
+from .auth_service import auth_service
 
 __all__ = [
-    # TODO: Add service exports as they are implemented
```
Comment (aguilar1x) [inline app/services/__init__.py:20]: don´t remove this comment pls

File: app/models/users/__init__.py [hunk 9]
```diff
@@ -2,16 +2,18 @@
...
-# from .user import User
-# from .profile import UserProfile
+from .user import User, UserType
+
+from sqlalchemy.dialects.postgresql import ENUM
+
+usertype_enum = ENUM(UserType, name="usertype")
 
 __all__ = [
-    # TODO: Add user model exports as they are implemented
```
Comment (aguilar1x) [inline app/models/users/__init__.py:16]: don´t remove this comment

File: app/services/auth_service.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                token,
+                self.settings.secret_key,
+                algorithms=[self.settings.algorithm]
+            )
+            email = payload.get("sub")
+            if email is None:
+                raise credentials_exception
+            token_data = TokenData(email=email)
+        except JWTError:
+            raise credentials_exception
```
[the remark this answers is not in the document]
Comment (aguilar1x) [reply app/services/auth_service.py:123]: Fix from none for better debugging

**Comment by WassCodeur** [comment]
Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501). 

Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?

**Comment by aguilar1x** [comment]
> Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501).
> 
> Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?

Ready, they are no longer limited to 77. You can normally review lint errors with the "make lint" command

**Comment by aguilar1x** [comment]
> > > Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501).
> > > Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?
> > 
> > 
> > Ready, they are no longer limited to 77. You can normally review lint errors with the "make lint" command
> 
> Thank you

Let me know when the final changes and fixes are complete.

**Comment by WassCodeur** [comment]
> > > > Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501).
> > > > Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?
> > > 
> > > 
> > > Ready, they are no longer limited to 77. You can normally review lint errors with the "make lint" command
> > 
> > 
> > Thank you
> 
> Let me know when the final changes and fixes are complete.

The maximum length should be 88 and 79 as this is what is allowed with `black`.

Screenshot from 2025-07-05 02-30-09

**Comment by aguilar1x** [comment]
> > > > > Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501).
> > > > > Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?
> > > > 
> > > > 
> > > > Ready, they are no longer limited to 77. You can normally review lint errors with the "make lint" command
> > > 
> > > 
> > > Thank you
> > 
> > 
> > Let me know when the final changes and fixes are complete.
> 
> The maximum length should be 88 and 79 as this is what is allowed with `black`.
> 
> Screenshot from 2025-07-05 02-30-09

That's because you haven't pulled the changes from main.

Add this file, which is the one that resolves those errors.

Screenshot 2025-07-04 at 9 23 21 PM

**Comment by WassCodeur** [comment]
> > > > > > Hi @aguilar1x , I noticed a formatting conflict between `flake8` and `black` regarding line length (E501).
> > > > > > Would you be okay with ignoring E501 in `flake8` to align with `black`’s defaults (88 chars), or should I manually keep lines within the limit ?
> > > > > 
> > > > > 
> > > > > Ready, they are no longer limited to 77. You can normally review lint errors with the "make lint" command
> > > > 
> > > > 
> > > > Thank you
> > > 
> > > 
> > > Let me know when the final changes and fixes are complete.
> > 
> > 
> > The maximum length should be 88 and 79 as this is what is allowed with `black`.
> > Screenshot from 2025-07-05 02-30-09
> 
> That's because you haven't pulled the changes from main.
> 
> Add this file, which is the one that resolves those errors.
> 
> Screenshot 2025-07-04 at 9 23 21 PM

Now it's okay

**Comment by aguilar1x** [comment]
> Screenshot from 2025-07-05 12-40-26
> 
> Hello @aguilar1x
> 
> This PR focuses on Implementing user authentication features including registration, login, and JWT token management.
> 
> While running mypy, I noticed several type-related issues across the project (e.g., missing type annotations like **all**, untyped functions, etc.).
> 
> To avoid mixing concerns and overloading this PR, I would like to propose the following:
> 
> Open a separate issue dedicated to adding type hints and improving static type checking (mypy) Then submit a new PR specifically focused on applying those improvements This will help keep this PR focused and maintain readability while progressively improving the codebase. Would you be okay with that approach? Thanks in advance!

gm,

Ah yes, I'm already aware of those, but that will be left as a refactor for a new issue.

Thank you very much for your contribution to our project. I hope you enjoyed collaborating with us and building alongside our team.

Keep building. 🏗️

**Review by aguilar1x — APPROVED** [review]
LGTM!

Keep building.

**Comment by WassCodeur** [comment]
> > Screenshot from 2025-07-05 12-40-26
> > Hello @aguilar1x
> > This PR focuses on Implementing user authentication features including registration, login, and JWT token management.
> > While running mypy, I noticed several type-related issues across the project (e.g., missing type annotations like **all**, untyped functions, etc.).
> > To avoid mixing concerns and overloading this PR, I would like to propose the following:
> > Open a separate issue dedicated to adding type hints and improving static type checking (mypy) Then submit a new PR specifically focused on applying those improvements This will help keep this PR focused and maintain readability while progressively improving the codebase. Would you be okay with that approach? Thanks in advance!
> 
> gm,
> 
> Ah yes, I'm already aware of those, but that will be left as a refactor for a new issue.
> 
> Thank you very much for your contribution to our project. I hope you enjoyed collaborating with us and building alongside our team.
> 
> Keep building. 🏗️

Thank you very much, I enjoyed working with you and especially thank you for your patience.

I definitely fell in love with the project and I'm going to stay here.
